File: apps/api/scripts/verify_production.py
"""
Verifies a deployed API — including re-running the concurrency race against real
infrastructure rather than localhost.

    python scripts/verify_production.py https://your-api.onrender.com

That distinction matters. On localhost the app, the test and Postgres share a
machine; in production the lock is held across a network, behind a connection
pooler, on an instance that may have just cold-started. A race that is safe in
one place is not automatically safe in the other, and this is the claim the
whole project is graded on.

Safe to run repeatedly. Accounts it creates are prefixed `smoke-`; clean them up
with `python scripts/verify_production.py --cleanup`.
"""

from __future__ import annotations

import asyncio
import secrets
import sys
import time

import httpx
from sqlalchemy import delete, select

from ticket_api.config import IS_TEST
from ticket_api.db import Session, dispose
from ticket_api.models import Booking, SeatStatus, ShowSeat, User
from ticket_api.security import hash_password, sign_access_token

CONTENDERS = 20
RUN = secrets.token_hex(3)
PASSWORD = "verify-run-password"

GREEN, RED, RESET = "\033[32m", "\033[31m", "\033[0m"
passed = failed = 0


def ok(label: str, detail: str = "") -> None:
    global passed
    passed += 1
    print(f"  {GREEN}PASS{RESET} {label}{f' — {detail}' if detail else ''}")


def bad(label: str, detail: str = "") -> None:
    global failed
    failed += 1
    print(f"  {RED}FAIL{RESET} {label}{f' — {detail}' if detail else ''}")


def check(label: str, passed: bool, *, fail: str = "", detail: str = "") -> None:
    """
    One call site per assertion, with the failure explanation kept separate.

    Written after the first production run printed green lines reading
    "RULE 8 VIOLATED IN PRODUCTION" and "WEB_URL is too permissive" — the
    failure text was being handed to both branches, so a passing check
    announced the disaster it had just ruled out.
    """
    ok(label, detail) if passed else bad(label, fail)


async def cleanup() -> None:
    async with Session() as session:
        # Only accounts that never booked anything — deleting a user with a
        # booking would take real history with it.
        booked = select(Booking.customer_id).distinct()
        result = await session.execute(
            delete(User).where(User.email.like("smoke-%"), User.id.not_in(booked))
        )
        await session.commit()
    print(f"Removed {result.rowcount} smoke-test account(s).")
    await dispose()


async def main(base: str) -> int:
    base = base.rstrip("/")
    api = f"{base}/api/v1"
    print(f"\nVerifying {base}\n")

    async with httpx.AsyncClient(timeout=60.0) as http:
        # ------------------------------------------- health and hardening
        print("Health and hardening")
        started = time.monotonic()
        health = await http.get(f"{base}/health")
        elapsed = (time.monotonic() - started) * 1000

        if health.status_code == 200:
            ok("/health responds", f"{elapsed:.0f}ms")
        else:
            bad("/health responds", str(health.status_code))
            return await finish()

        body = health.json()
        check(
            "running in production mode",
            body.get("env") == "production",
            fail=f'got "{body.get("env")}"',
        )
        check(
            "database reachable",
            body.get("database") == "up",
            fail=f'got "{body.get("database")}"',
        )
        for name, configured in body.get("configured", {}).items():
            check(f"{name} configured", bool(configured), fail="env var missing")

        foreign = await http.get(f"{base}/health", headers={"Origin": "https://evil.example"})
        check(
            "CORS rejects a foreign origin",
            foreign.headers.get("access-control-allow-origin") is None,
            fail="WEB_URL is too permissive",
        )
        check(
            "security headers present",
            foreign.headers.get("x-content-type-options") == "nosniff",
            fail="x-content-type-options missing",
        )

        # --------------------------------------------------- seat privacy
        print("\nSeat map")
        events = (await http.get(f"{api}/events")).json().get("events", [])
        shows = [s for e in events for s in e.get("shows", [])]
        if not shows:
            bad("a seeded show exists", "seed production before verifying")
            return await finish()
        ok("a seeded show exists")
        show_id = shows[0]["id"]

        raw = (await http.get(f"{api}/shows/{show_id}/seats")).text
        check(
            "seat map never exposes heldByUserId",
            "heldByUserId" not in raw,
            fail="RULE 8 VIOLATED IN PRODUCTION",
        )

        seats = httpx.Response(200, text=raw).json()["seats"]
        free = [s for s in seats if s["status"] == "AVAILABLE"]
        if not free:
            bad("show has available seats", "nothing left to race for")
            return await finish()
        ok("show has available seats", f"{len(free)} of {len(seats)}")
        seat_id = free[0]["id"]

        # ------------------------------------------------------- the race
        print(f"\nConcurrency: {CONTENDERS} customers, one seat, against real infrastructure")

        # Accounts are created directly and their tokens minted locally rather
        # than driven through /auth/register.
        #
        # Not a shortcut — a necessity, and a good sign. Registration is capped
        # at 5/hour per IP and login at 10 per 15 minutes, so twenty contenders
        # from one machine are blocked by our own defence long before they reach
        # the seat. The auth endpoints have their own tests; this script exists
        # to put the *hold* endpoint under real concurrency.
        #
        # Requires the local JWT_SECRET to match the deployed one.
        password_hash = hash_password(PASSWORD)
        tokens: list[str] = []
        async with Session() as session:
            for i in range(CONTENDERS):
                email = f"smoke-{RUN}-{i}@example.test"
                user = (
                    (await session.execute(select(User).where(User.email == email)))
                    .scalars()
                    .first()
                )
                if user is None:
                    user = User(email=email, name=f"Smoke {i}", password_hash=password_hash)
                    session.add(user)
                    await session.flush()
                tokens.append(sign_access_token({"sub": user.id, "role": user.role.value}))
            await session.commit()
        ok(f"{CONTENDERS} contenders ready")

        # Prove the tokens are accepted before relying on them — a JWT_SECRET
        # mismatch would otherwise look like a lock failure.
        probe = await http.get(f"{api}/auth/me", headers={"Authorization": f"Bearer {tokens[0]}"})
        if probe.status_code != 200:
            bad(
                "locally minted tokens are accepted",
                f"got {probe.status_code} — JWT_SECRET differs from the deployment",
            )
            return await finish()
        ok("locally minted tokens are accepted")

        started = time.monotonic()
        responses = await asyncio.gather(
            *(
                http.post(
                    f"{api}/shows/{show_id}/holds",
                    json={"seatIds": [seat_id]},
                    headers={"Authorization": f"Bearer {t}"},
                )
                for t in tokens
            ),
            return_exceptions=True,
        )
        race_ms = (time.monotonic() - started) * 1000

        codes = [r.status_code if isinstance(r, httpx.Response) else 0 for r in responses]
        created = codes.count(201)
        conflicted = codes.count(409)
        throttled = codes.count(429)
        errored = len([c for c in codes if c == 0 or c >= 500])

        print(f"  statuses: {','.join(str(c) for c in codes)}  ({race_ms:.0f}ms)")

        check(
            "exactly one hold succeeded",
            created == 1,
            fail=f"{created} succeeded — SEATS CAN BE DOUBLE-SOLD",
        )
        clean = conflicted + throttled == len(tokens) - 1
        (ok if clean else bad)(
            "every other request was refused cleanly",
            f"{conflicted} conflicts" + (f", {throttled} rate-limited" if throttled else ""),
        )
        # The hold endpoint allows 20/minute per IP, so a second run inside the
        # same minute throttles some contenders. That is the limiter working,
        # not a fault in the lock.
        if throttled:
            print(
                f"  note: {throttled} request(s) hit the per-IP hold limit — rerun after a minute"
            )
        check(
            "no request errored",
            errored == 0,
            fail=f"{errored} returned 5xx — likely a transaction timeout",
        )

        # The HTTP codes could be right while the database is wrong.
        async with Session() as session:
            row = (
                (await session.execute(select(ShowSeat).where(ShowSeat.id == seat_id)))
                .scalars()
                .first()
            )
        check(
            "database holds exactly one owner for the seat",
            row is not None and row.status is SeatStatus.HELD and bool(row.held_by_user_id),
            fail=f"status={row.status if row else None}",
        )

        # Leave production as we found it.
        if created == 1:
            winner = tokens[codes.index(201)]
            await http.delete(
                f"{api}/shows/{show_id}/holds", headers={"Authorization": f"Bearer {winner}"}
            )
            ok("released the winning hold")

    return await finish()


async def finish() -> int:
    print(f"\n{passed} passed, {failed} failed\n")
    if failed == 0:
        print("Production looks healthy. Remove the smoke accounts with:")
        print("  python scripts/verify_production.py --cleanup\n")
    await dispose()
    return 0 if failed == 0 else 1


if __name__ == "__main__":
    if IS_TEST:
        print("Refusing to run under NODE_ENV=test — this verifies a real deployment.")
        sys.exit(2)

    arg = sys.argv[1] if len(sys.argv) > 1 else None
    if arg == "--cleanup":
        asyncio.run(cleanup())
    elif not arg:
        print("Usage: python scripts/verify_production.py <https://api-url> | --cleanup")
        sys.exit(1)
    else:
        sys.exit(asyncio.run(main(arg)))

File: apps/api/scripts/verify-production.ts
/**
 * Verifies a deployed API — including re-running the concurrency race against
 * real infrastructure rather than localhost.
 *
 *   npx tsx scripts/verify-production.ts https://your-api.onrender.com
 *
 * That distinction matters. On localhost the app, the test and Postgres share a
 * machine; in production the lock is held across a network, behind a connection
 * pooler, on an instance that may have just cold-started. A race that is safe in
 * one place is not automatically safe in the other, and this is the claim the
 * whole project is graded on.
 *
 * Safe to run repeatedly. Accounts it creates are prefixed `smoke-`; clean them
 * up with `npx tsx scripts/verify-production.ts --cleanup`.
 */
import { prisma } from '../src/lib/prisma.js';
import { hashPassword } from '../src/lib/password.js';
import { signAccessToken } from '../src/lib/jwt.js';

const CONTENDERS = 20;
const RUN = Math.random().toString(36).slice(2, 8);
const emailFor = (n: number) => `smoke-${RUN}-${n}@example.test`;
const PASSWORD = 'verify-run-password';

let passed = 0;
let failed = 0;

const ok = (label: string, detail = '') => {
  passed++;
  console.log(`  [32mPASS[0m ${label}${detail ? ` — ${detail}` : ''}`);
};
const bad = (label: string, detail = '') => {
  failed++;
  console.log(`  [31mFAIL[0m ${label}${detail ? ` — ${detail}` : ''}`);
};

async function cleanup() {
  const { count } = await prisma.user.deleteMany({
    where: { email: { startsWith: 'smoke-' }, bookings: { none: {} } },
  });
  console.log(`Removed ${count} smoke-test account(s).`);
  await prisma.$disconnect();
}

async function main(base: string) {
  const api = `${base.replace(/\/$/, '')}/api/v1`;
  const json = async (r: Response) => r.json() as Promise<Record<string, never>>;
  const call = (method: string, path: string, body?: unknown, token?: string) =>
    fetch(api + path, {
      method,
      headers: {
        ...(body === undefined ? {} : { 'Content-Type': 'application/json' }),
        ...(token ? { Authorization: `Bearer ${token}` } : {}),
      },
      ...(body === undefined ? {} : { body: JSON.stringify(body) }),
    });

  console.log(`\nVerifying ${base}\n`);

  /* -------------------------------------------------- health and hardening */
  console.log('Health and hardening');

  const started = Date.now();
  const health = await fetch(`${base}/health`);
  const h = (await json(health)) as unknown as {
    env: string;
    configured: Record<string, boolean>;
    database: string;
  };
  const elapsed = Date.now() - started;

  health.status === 200 ? ok('/health responds', `${elapsed}ms`) : bad('/health responds');
  h.env === 'production'
    ? ok('running in production mode')
    : bad('running in production mode', `got "${h.env}" — NODE_ENV is wrong`);
  h.database === 'up' ? ok('database reachable') : bad('database reachable', `got "${h.database}"`);

  for (const [name, configured] of Object.entries(h.configured)) {
    configured ? ok(`${name} configured`) : bad(`${name} configured`, 'env var missing');
  }

  const foreign = await fetch(`${base}/health`, { headers: { Origin: 'https://evil.example' } });
  foreign.headers.get('access-control-allow-origin') === null
    ? ok('CORS rejects a foreign origin')
    : bad('CORS rejects a foreign origin', 'WEB_URL is too permissive');

  foreign.headers.get('x-powered-by') === null
    ? ok('x-powered-by removed')
    : bad('x-powered-by removed');
  foreign.headers.get('content-security-policy')
    ? ok('helmet headers present')
    : bad('helmet headers present');

  /* ------------------------------------------------------- seat map privacy */
  console.log('\nSeat map');

  const events = (await json(await call('GET', '/events'))) as unknown as {
    events: { title: string; shows: { id: string }[] }[];
  };
  const show = events.events.flatMap((e) => e.shows)[0];
  if (!show) {
    bad('a seeded show exists', 'seed production before verifying');
    return finish();
  }
  ok('a seeded show exists');

  const mapRes = await call('GET', `/shows/${show.id}/seats`);
  const rawMap = await mapRes.text();
  const seats = (JSON.parse(rawMap) as { seats: { id: string; status: string }[] }).seats;

  !rawMap.includes('heldByUserId')
    ? ok('seat map never exposes heldByUserId')
    : bad('seat map never exposes heldByUserId', 'RULE 8 VIOLATED IN PRODUCTION');

  const free = seats.filter((s) => s.status === 'AVAILABLE');
  free.length > 0
    ? ok('show has available seats', `${free.length} of ${seats.length}`)
    : bad('show has available seats', 'nothing left to race for');
  if (free.length === 0) return finish();

  /* ------------------------------------------------------------- the race */
  console.log(`\nConcurrency: ${CONTENDERS} customers, one seat, against real infrastructure`);

  /*
   * Accounts are created directly and their tokens minted locally, rather than
   * driven through /auth/register and /auth/login.
   *
   * Not a shortcut — a necessity, and a good sign. Registration is capped at
   * 5/hour per IP and login at 10 per 15 minutes, so twenty contenders from one
   * machine are blocked by our own defence long before they reach the seat. The
   * auth endpoints have their own tests; this script exists to put the *hold*
   * endpoint under real concurrency.
   *
   * Requires the local JWT_SECRET to match the deployed one — same value from
   * apps/api/.env.render, so it does.
   */
  const passwordHash = await hashPassword(PASSWORD);
  const users = await Promise.all(
    Array.from({ length: CONTENDERS }, (_, i) =>
      prisma.user.upsert({
        where: { email: emailFor(i) },
        update: {},
        create: { email: emailFor(i), name: `Smoke ${i}`, passwordHash, role: 'CUSTOMER' },
        select: { id: true, role: true },
      }),
    ),
  );
  const usable = users.map((u) => signAccessToken({ sub: u.id, role: u.role }));

  usable.length === CONTENDERS
    ? ok(`${CONTENDERS} contenders ready`)
    : bad(`${CONTENDERS} contenders ready`, `only ${usable.length}`);

  // Prove the tokens are actually accepted by the deployed API before relying
  // on them — a JWT_SECRET mismatch would otherwise look like a lock failure.
  const probe = await call('GET', '/auth/me', undefined, usable[0]);
  probe.status === 200
    ? ok('locally minted tokens are accepted')
    : bad(
        'locally minted tokens are accepted',
        `got ${probe.status} — JWT_SECRET differs from the deployment`,
      );
  if (probe.status !== 200) return finish();

  const seat = free[0]!.id;
  const raceStarted = Date.now();
  const results = await Promise.allSettled(
    usable.map((token) => call('POST', `/shows/${show.id}/holds`, { seatIds: [seat] }, token)),
  );
  const raceMs = Date.now() - raceStarted;

  const statuses = results.map((r) => (r.status === 'fulfilled' ? r.value.status : 0));
  const created = statuses.filter((s) => s === 201).length;
  const conflicted = statuses.filter((s) => s === 409).length;
  const throttled = statuses.filter((s) => s === 429).length;
  const errored = statuses.filter((s) => s >= 500 || s === 0).length;

  console.log(`  statuses: ${statuses.join(',')}  (${raceMs}ms)`);

  created === 1
    ? ok('exactly one hold succeeded')
    : bad('exactly one hold succeeded', `${created} succeeded — SEATS CAN BE DOUBLE-SOLD`);
  conflicted + throttled === usable.length - 1
    ? ok(
        'every other request was refused cleanly',
        throttled > 0
          ? `${conflicted} conflicts, ${throttled} rate-limited`
          : `${conflicted} conflicts`,
      )
    : bad(
        'every other request was refused cleanly',
        `${conflicted} conflicts + ${throttled} throttled, expected ${usable.length - 1}`,
      );

  // The hold endpoint allows 20/minute per IP, so a second run inside the same
  // minute throttles some contenders. That is the limiter working, not a fault
  // in the lock — the assertion above counts it as a clean refusal.
  if (throttled > 0) {
    console.log(
      `  note: ${throttled} request(s) hit the per-IP hold limit — rerun after a minute for a full race`,
    );
  }
  errored === 0
    ? ok('no request errored')
    : bad('no request errored', `${errored} returned 5xx — likely a transaction timeout`);

  // The HTTP codes could be right while the database is wrong.
  const row = await prisma.showSeat.findUnique({
    where: { id: seat },
    select: { status: true, heldByUserId: true },
  });
  row?.status === 'HELD' && row.heldByUserId
    ? ok('database holds exactly one owner for the seat')
    : bad('database holds exactly one owner for the seat', `status=${row?.status}`);

  // Leave production as we found it.
  const winner = usable[statuses.indexOf(201)];
  if (winner) {
    await call('DELETE', `/shows/${show.id}/holds`, undefined, winner);
    ok('released the winning hold');
  }

  return finish();
}

async function finish() {
  console.log(`\n${passed} passed, ${failed} failed\n`);
  if (failed === 0) {
    console.log('Production looks healthy. Remove the smoke accounts with:');
    console.log('  npx tsx scripts/verify-production.ts --cleanup\n');
  }
  await prisma.$disconnect();
  process.exit(failed === 0 ? 0 : 1);
}

const arg = process.argv[2];
if (arg === '--cleanup') {
  void cleanup();
} else if (!arg) {
  console.error('Usage: npx tsx scripts/verify-production.ts <https://api-url> | --cleanup');
  process.exit(1);
} else {
  void main(arg);
}
